scipy/optimize/_differentiable_functions.py

```diff
@@ -145,11 +151,17 @@ def update_grad():
                 self.g = grad_wrapped(self.x)
 
         elif grad in FD_METHODS:
+            lb, ub = _prepare_bounds(finite_diff_bounds, self.x)
+            equal_bounds = lb == ub
+
             def update_grad():
                 self._update_fun()
                 self.ngev += 1
                 self.g = approx_derivative(fun_wrapped, self.x, f0=self.f,
                                            **finite_diff_options)
+                # when the lower and upper bounds are equal then set the
+                # gradient to 0.
+                self.g[equal_bounds] = 0.0
```

Comment:
This is the core of the PR: set to 0 derivatives of fixed parameters (which were Nans). I wish it is back-ported to the 1.5 series.

Reply:
It seems that with the change in `approx_derivative`, `self.g[equal_bounds]` will be filled with zeros by `approx_derivative`. Filling it explicitly is fine too.

Reply:
A more important question: could someone explain qualitatively why this change is fundamentally correct?

I don't want to crash into this PR all of the sudden and destroy all the work, but I have some concerns.

1. From a logical/mathematical point of view, the fact that `lb[i] == ub[i]` doesn't impose any constraints on the gradient at `lb[i]`. So it's very unclear what setting elements of `g` to zero achieves.
2. What is going to happen if we run the same problem now with analytical and numerical differentiation and some bounds being equal? In principle we must get close and correct results (including `x_opt[i] == lb[i] == ub[i]`), but I don't see how it is guaranteed with the current code.
3. It seems to me that the ability of a solver to handle `lb[i] == ub[i]` correctly with all possible settings requires more considerations and testing.

Reply:
I'm less concerned about losing work for the PR, and more concerned that we get behaviour right here. It'd be great to have more input with people familiar with optimisation, to make sure that happens. @nmayorov I'd appreciate any further input you're able to give.

The changes in this PR are by no means perfect.

1. Logically setting `lb[i] == ub[i]` means that you know, with 100% certainty, that `x[i]` has to be a constant value. Should a function have a gradient for `x[i]` if it's not allowed to take any other value? A graph of f vs x would show a singular dot, and not a line. In such a case the gradient should probably be undefined (nan), even for an 'analytic' derivative (or should it be zero?).

2. An analytic jacobian function (derived for `lb[i] < ub[i]`) may be able to calculate the gradient at `x[i] == lb[i]` with `lb[i] == ub[i]`. However, estimating the gradient with finite differences at that point *and respecting bounds* is impossible. If bounds are ignored during numeric differentiation, then it is possible to do that (you're also saying that you're not 100% sure that `lb[i] == ub[i]`). 
There were two sets of issues reported that have been involved in the lead to this PR:

- (a) estimating grad via finite differences should respect bounds. This lead to the use of `approx_derivative` that does so.
- (b) when the bounds-respecting was done for the previous issue this lead to issues in numerical optimizers, such as `slsqp` or `lbfgsb` that don't work when the jacobian contains `nan` (`df/dx=0.0/0.0`)

From testing SLSQP and LBFGSB do work when those nan are replaced by zeros, so that was the approach I took here. A different approach is to recognise problems with equal bounds and factor out those fixed parameters, thereby reducing the dimensionality of the problem. I think this factorisation would increase code complication and not straightforward to implement across all the solvers.
Yet another approach could be to get the actual iterative code in SLSQP and LBFGSB to behave properly for `lb[i] == ub[i]`, I don't think anyone has the appetite for altering Fortran code.

I don't think that this PR, or an alternate that factored out fixed parameters, would lead to the analytic or numerical differentation approaches giving the same answer unless we rolled back the changes from 2a.

TLDR: even though it's not a great solution, it may be the best of a bad lot.

Reply:
If one of the stumbling blocks is the alteration of nan-->0 in `_differentiable_functions`, then that could be done in a wrapper function at a higher level.

Reply:
I understand that you want to fix a problem with optimization with equal bounds failing due to valid restrictions of numerical differentiation.

I think we can split the problem into 2 parts:

1. Determine which solvers support equal bounds correctly. If the solver doesn't support it -- we should raise an error immediately in all cases.
2. If the solver does support equal bounds and the problem is formulated with equal bounds and numerical differentiation is requested then we have 3 options: raise an appropriate error, remove restrictions on allowed steps in `approx_derivative` or try the trick with zero gradient proposed here. 

I'll try to think some more about the situation and investigate how existing solvers supposed to work with fixed variables.

----

And I actually like the idea of automatic pre-elimination of fixed variables, it seems like a universal and correct approach. Not impossible to implement too.

Reply:
Elimination has a few issues. User functions have to be sent the full array, but the solvers have to be provided with a reduced set. So one has to write a wrapper for the actual problem, but also wrappers for any constraints machinery (e.g. slsqp). Finally one has to expand what the (fortran) solver returns so that OptimizeResult contains the correct number of variables.
i think there is large amount of tinkering required for each minimizer method, I don't think it's possible at a higher level.

Reply:
Not 100% sure, but conceptually it looks feasible using Python-side wrappers. The problem must be completely reformulated  to remove fixed variable, then solved by the optimizer, then the result is restored to contain the fixed variables.

Reply:
@nmayorov, the elimination solution is likely to take some time, and I think it's vital to get at least a temporary fix in for the next release. At the moment this is either going to be this PR, or #13069. The alternate PR lifts the issue from the numdiff code up into the solver method code.

Reply:
Are you sure it's vital to get such an ad-hoc fix? I mean if it fails with an error - it's an expected behavior and the error message is correct in this case. The case with equal bounds seems to be  "undefined area" in the current state to be honest. 

Well I'm not actively involved in the development as of late, so don't want to block anything. The second PR looks preferable to me.

As for `_approx_derivative` - maybe it should even raise an error with equal bounds, because it's incorrect setting considering the logic involved. But it won't play well with you changes, so it just a thought.

Reply:
The reported issues (+ their volume) and the [documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.Bounds.html#scipy.optimize.Bounds) for minimize indicate that setting equal bounds has historically been acceptable/valid.
However, the last release made the numerical differentiation obey any bounds, so back compatibility with this historical behaviour was broken (because the core of slsqp+lbfgsb can't cope with NaN in the gradient). Squash one bug, and another pops up.
Given that the alternate approach in #13069 might be preferable, would you have time to review that PR? It would be so, so, helpful. I agree that the change of changing Nan entries to zero in the gradient isn't 100% desirable. However, for the test example I constructed the modifications worked, they also work for various users posting these issues.

Reply:
@andyfaff @nmayorov do you see problems with the approach in gh-13096 (different from gh-13069)?